### ISSUE TITLE
Can someone explain to me how getNewCoordsForTrain works

### ISSUE THREAD

**Issue by davidwengier** [jamesformica/Trains#19]
I can't remember :(

Why does it pass in the speed? And then slowly reduce the speed? It adjusts the coords in pieces, which I guess makes sense to make it less jerky.. but it doesn't do that to the carriages at the same time???

Thanks!

**Comment by ScottHolden** [comment]
Speed is really the wrong name, from memory it is 'distance to move', so given a 'distance to move' it travels as far as it can in one cell, and then returns the remaining speed, aka, the 'remaining distance to move'. I think that's then repeated untill remaining speed is zero.
(All for memory and a quick peek, so could be 100% wrong)

**Comment by davidwengier** [comment]
Yeah, thats what it looks like.. but since the loop happens before moving that carriages, I would expect the trains and carriages to look like an accordion, but they don't, so I'm wondering if its actually needed.

**Comment by ScottHolden** [comment]
I think the carriages follow the same logic in their own loop

**Comment by davidwengier** [comment]
Yes, they do, but after the train has finished its loop.

Maybe the speeds just aren't big enough for the accordion-ing to be noticeable, and at high enough speeds its too fast to see ¯\\\_(ツ)_/¯

**Comment by ScottHolden** [comment]
Sounds about right :)

You could try breakpointing at the start of the carriage move loop to test. 

I don't think we are drawing to a hidden canvas and flipping, so at a high speed (or slow fps), then I guess you would get tearing.

**Comment by ScottHolden** [comment]
Oh, totally forgot that game loop and draw loop are seperate, that could lead to some interesting race conditions...
Though hopefully the movement should complete super super fast.

**Comment by davidwengier** [comment]
Everything is pretty fast. Plus the carriages are stored as a linked list, so good practice for you know what 😛

**Comment by jamesformica** [comment]
Haha I am thrilled the community behind Trains is still thriving! What are you up to? Building trains 2.0?

P.S do we still have awesome function names? I think i remember there being 'stopHammerTime'

**Comment by alexkeil** [comment]
This is still a thing? 😂 that's awesome!

**Comment by davidwengier** [comment]
Yep, still awesome function names for the most part, though I have to admit it does make revisiting the code difficult.

I forked it to https://github.com/davidwengier/trains2 because I wanted to make some changes, but cbf setting up Ruby etc. so its just typescript and html. Happy to contribute back to master if you want.

We need to add credits to the front page too, every time anyone posts about it on Twitter it always gets a good response.

**Comment by ScottHolden** [comment]
So, do we contribute here, or trains2?

**Comment by davidwengier** [comment]
Up to @jamesformica as to whether he wants me to PR my changes into this?

**Comment by davidwengier** [comment]
Unless you want to add in explosions! Then I'll stick to my fork :P

**Comment by alexkeil** [comment]
Well I put in collisions and you all didnt like them haha

**Comment by ScottHolden** [comment]
Nah, not explosions, just to fix the drawing math on the corner tracks, it just doesn't feel right :/

**Comment by davidwengier** [comment]
There are issues.. if you go to https://wengier.com/trains2/play.html?debug you can see that sometimes when a train has carriages, some tracks in the middle aren't occupied. The hittesting also doesn't care about track connections, so sometimes some trains wait on corners etc.

Would love to do deadlock detection too, perhaps with (really really slow) back off.

**Comment by jamesformica** [comment]
No no all new work should be done on @davidwengier's repo! This is a time capsule to always be able to look back at what we created!

Haha i was mega confused because i remember making a front-end only version of trains so I could host it on github pages because having ruby was redundant anyway.. Then it clicked, I have a branch gh-pages which is full front end lol

If there is anything I can help with on the new version let me know 😄

**Comment by davidwengier** [comment]
Thanks James <3

I started from your gh-pages branch actually, as was just going to muck around there, but the experience with typescript was so much better than javascript I couldn't do it. Should probably set up some proper build at some point.

Anyone who wants to, feel free to issue and/or PR in trains2.
